src/frontend/context/SecurityContext.tsx

```diff
@@ -0,0 +1,128 @@
+import * as React from "react";
+import AsyncStorage from "@react-native-async-storage/async-storage";
+import { ASYNC_KILL_KEY, PASSWORD_KEY } from "../constants";
+import { useEffect } from "react";
+import { useRef } from "react";
+
+//We should be setting the password to null when the password is turned off
+interface SecuritContextType {
+  killState: boolean;
+  setKillState: (killStateValue: boolean) => void;
+  passcode: string | null;
+  setPasscode: (passValue: string | null) => void;
+  checkFlag: boolean;
+  killStateActive: boolean;
+  setKillStateActive: (isActive: boolean) => void;
+  setCheckFlag: (checkFlag: boolean) => void;
+}
+
+const DEFAULT_CONTEXT: SecuritContextType = {
+  killState: false,
+  setKillState: () => {},
+  setPasscode: () => {},
+  passcode: null,
+  checkFlag: false,
+  setCheckFlag: () => {},
+  killStateActive: false,
+  setKillStateActive: () => {},
+};
+
+export const SecurityContext = React.createContext<SecuritContextType>(
+  DEFAULT_CONTEXT
+);
+
+export const SecurityProvider = ({
+  children,
+}: {
+  children: React.ReactNode;
+}) => {
+  const [killState, setKillState] = React.useState(false);
+  const [killStateActive, setKillStateActive] = React.useState(false);
+  const [passcode, setPasscode] = React.useState<string | null>(null);
+  const [checkFlag, setCheckFlag] = React.useState(false);
+  const firstRender = useRef(true);
+
+  useEffect(() => {
+    if (firstRender.current) {
+      AsyncStorage.getItem(PASSWORD_KEY, (err, val) => {
+        if (!!val) {
+          setPasscode(val);
+          setCheckFlag(true);
+          AsyncStorage.getItem(ASYNC_KILL_KEY, (err, val) => {
+            if (!!val) setKillStateActive(val === "true");
+          });
+        } else {
+          firstRender.current = false;
+        }
+      });
+    }
+  }, []);
+
+  useEffect(() => {
+    //Checks if the pass is just being set from initial render
+    if (firstRender.current) {
+      firstRender.current = false;
+      return;
+    }
+
+    if (!passcode) {
+      setKillState(false);
+      setKillStateActive(false);
+      setCheckFlag(false);
+      AsyncStorage.setItem(PASSWORD_KEY, "");
+    } else {
+      AsyncStorage.setItem(PASSWORD_KEY, passcode);
```

Comment:
although it makes sense to run this in an effect since it's dependent on `passcode`, wondering if it's safer and less prone to catastrophic runaway effects if you call setItem in the passcode updater function `setPasscodeWithCheck`. just thinking that it could simplify the implementation a bit and remove the need for this effect, which is doing a somewhat a unideal check based on react render cycles

Reply:
I thought of doing that orignally, but I do want to update these BASED on an updated passcode. So this gurantees that when passcode is updated then asyncStorage would be effected

Reply:
that's reasonable. perhaps the async storage calls here can be moved to `setPasscodeWithCheck` since that's the only source of updates for the passcode state (aside from the initial render effect in this component). i like the idea of co-locating those operations to avoid indirection and potentially undesired updates to async storage 

think the firstRender check is kind of a code smell though so just wondering if there's a way to avoid it in this case